[3.3] Add pagination to datasetmanager.php

Atticus29 opened the pull request:
# Description

This PR closes #2200 . It also autoformats the invovled files quite a bit, so I added comments to the "Files changed" tab above where the meaningful changes were made.

Currently, the maximum number of records displayed per page is an arbitrary 200, but I'm happy to change it.

# Before

<img width="1582" alt="Screenshot 2025-02-10 at 3 56 30 PM" src="https://github.com/user-attachments/assets/78ba54df-2cc2-4462-919c-a3c2b1cb473e" />


# After

<img width="1582" alt="Screenshot 2025-02-10 at 3 49 56 PM" src="https://github.com/user-attachments/assets/43cb5b16-84b2-4ff0-969a-069aa64d377c" />


# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [ ] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
    - N/A
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [ ] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
    - N/A
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [ ] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
    - I did not do this, but I did note in the files changed where meaningful code changes were made.
- [ ] Comment which GitHub issue(s), if any does this PR address
    - [https://github.com/BioKIC/Symbiota/issues/2200](https://github.com/BioKIC/Symbiota/issues/2200)
- [ ] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).
    - N/A

# Post-Approval

- [ ] It is the code author's responsibility to merge their own pull request after it has been approved
- [ ] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [ ] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [ ] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [ ] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [ ] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [ ] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [ ] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [ ] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [ ] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

Atticus29 committed:
add pagination to datasetmanager.php

Atticus29 committed:
add translations

Atticus29 commented on checklists/checklist.php:
```diff
@@ -1,14 +1,15 @@
 <?php
```
Just autoformatting on this page.

Atticus29 committed:
remove cruft

Atticus29 commented on classes/OccurrenceDataset.php:
```diff
@@ -1,51 +1,57 @@
 	private $collArr = array();
 	private $datasetId = 0;
 	private $errorArr = array();
+	private $occurrenceCount = null;
```
new line.

Atticus29 commented on classes/OccurrenceDataset.php:
```diff
@@ -121,208 +129,247 @@ public function getDatasetArr(){
 	}
 
-	public function getOccurrences($datasetId){
+	public function setOccurrenceCount($datasetId)
```
new method

Atticus29 commented on classes/OccurrenceDataset.php:
```diff
@@ -121,208 +129,247 @@ public function getDatasetArr(){
+		}
+	}
+
+	public function getOccurrences($datasetId, $pageNumber = 1, $retLimit = 500)
```
modified method

Atticus29 commented on classes/OccurrenceDataset.php:
```diff
@@ -121,208 +129,247 @@ public function getDatasetArr(){
+		return $returnVal;
+	}
+
+	public function getOccurrenceCount($datasetId)
```
new method

Atticus29 commented on collections/listtabledisplay.php:
```diff
@@ -1,18 +1,18 @@
 <?php
```
All autoformatting

Atticus29 committed:
checkout Development versions of checklists/checklist.php and collections/listtabledisplay.php

Atticus29 commented on collections/datasets/datasetmanager.php:
```diff
@@ -285,39 +278,57 @@ function targetDownloadPopup(f) {
-						if($occArr = $datasetManager->getOccurrences($datasetId)){
-							?>
+						$retLimit = 200;
+						if ($occArr = $datasetManager->getOccurrences($datasetId, $pageNumber, $retLimit)) {
```
Updated based on change in method signature.

Atticus29 committed:
remove more cruft and simplify

Atticus29 commented on collections/datasets/datasetmanager.php:
```diff
@@ -285,39 +278,55 @@ function targetDownloadPopup(f) {
-									<?php echo '<b>'.$LANG['COUNT'].': '.count($occArr).' '.$LANG['RECORDS'].'</b>'; ?>
+									<?php echo '<b>' . $LANG['COUNT'] . ': ' . count($occArr) . ' ' . $LANG['RECORDS'] . '</b>'; ?>
+								</div>
+								<div class="bottom-breathing-room-rel">
```
Add the pagination div

themerekat commented:
When you have lots of pages you get this monstrosity
![image](https://github.com/user-attachments/assets/a4c1a05a-e326-41b1-aa75-69c74480b790)

Can we hide a bunch of these under a dropdown list, or three dots in the middle then the last page is it's own link? Here's how we do it in collections/list
![image](https://github.com/user-attachments/assets/fd04fe25-b342-4e30-ad3b-a4294496c6c4)

Although I think I would rather have a dropdown list of all the in-between numbers and an actual number for the last page instead of just "Last".

themerekat reviewed (changes requested):
See previous comment

Atticus29 committed:
begin to implement jump to page in large paginators

Atticus29 committed:
move the start of the form

Atticus29 committed:
add styling to make one line and handle logic for which pages to display a little better

Atticus29 committed:
combine sliding window with ensuring end and beginning always show

Atticus29 committed:
add media queries and other styling changes

Atticus29 committed:
add responsive design

Atticus29 committed:
clean up cruft

Atticus29 commented:
> When you have lots of pages you get this monstrosity ![image](https://private-user-images.githubusercontent.com/16688784/412014409-a4c1a05a-e326-41b1-aa75-69c74480b790.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzNTk4MDgsIm5iZiI6MTczOTM1OTUwOCwicGF0aCI6Ii8xNjY4ODc4NC80MTIwMTQ0MDktYTRjMWEwNWEtZTMyNi00MWIxLWFhNzUtNjljNzQ0ODBiNzkwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDExMjUwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2ZWNlZDJkYjBmZDg1NzJjNDFjNmNmN2Q5NDg4ODA3MjQxY2JmMjM3NGQ3ZmEyN2EwZDIzYTRmMWI3ZmQ3ZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Z-dzgBWe5K1oDy9FIZM4z3SDMvHTImC3yrPKzRASBpg)
> 
> Can we hide a bunch of these under a dropdown list, or three dots in the middle then the last page is it's own link? Here's how we do it in collections/list ![image](https://private-user-images.githubusercontent.com/16688784/412015220-fd04fe25-b342-4e30-ad3b-a4294496c6c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzNTk4MDgsIm5iZiI6MTczOTM1OTUwOCwicGF0aCI6Ii8xNjY4ODc4NC80MTIwMTUyMjAtZmQwNGZlMjUtYjM0Mi00ZTMwLWFkM2ItYTQyOTQ0OTZjNmM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDExMjUwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3NWJhYTI3MTgyYmJjY2FkNmZjYzZiNjg2NWY4MTY3MGIwMmQ4ZWU1MTlmZDhlYzVhZTA2YjcyOGM0N2JjYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.z4TF8lKAPkssBxZKptjlnCJYj5UKCh4oWOAiMdkdeHY)
> 
> Although I think I would rather have a dropdown list of all the in-between numbers and an actual number for the last page instead of just "Last".

Should be resolved now. Per chat on Teams, links to neighboring 10 pages (when available; fewer if not; occasionally 11 because this logic was weird and hurt my brain) provided in a "sliding window" plus always a link to first page and last page. An input field for a page number is available in between for quick access to any target page. Some form validation present for said input field (users can't enter page options that don't exist). When the screen size gets too small to fit all of this, only the first page and last page links (as well a jump to page input field) are provided.

Atticus29 committed:
make slight logic improvement

themerekat commented:
@Atticus29  I found one edge case between 1050 and about 880 pixels where the last page floats off the page
![image](https://github.com/user-attachments/assets/05485280-c95b-4458-97ee-fa508c18c537)

Atticus29 committed:
add intermediate number of page links by getting rid of even page number links

Atticus29 committed:
handle first and last page links and extra | symbols

Atticus29 commented:
> @Atticus29 I found one edge case between 1050 and about 880 pixels where the last page floats off the page ![image](https://private-user-images.githubusercontent.com/16688784/412464838-05485280-c95b-4458-97ee-fa508c18c537.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzODQwMzEsIm5iZiI6MTczOTM4MzczMSwicGF0aCI6Ii8xNjY4ODc4NC80MTI0NjQ4MzgtMDU0ODUyODAtYzk1Yi00NDU4LTk3ZWUtZmE1MDhjMThjNTM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDE4MDg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlYmVmMDQ2MmM3M2YxOTQ2YjQ4YmUwM2IwMTM1MDgzMTliODI5YzEwN2NiNjkyODQwZjBjZDU4ZTllYWNiYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.772rJREoV8_2ORXFuK8KRIROKnff6tQiRefrNQfLAcQ)

Fixed in b2fc870c0 and 8cc5859b0. I added another behavior at intermediate screen sizes where only the odd page number links in the sliding window get shown.

themerekat reviewed (approved):
It's beautiful! Works locally.

MuchQuak commented on classes/OccurrenceDataset.php:
```diff
@@ -121,208 +129,247 @@ public function getDatasetArr(){
+		$returnVal = 0;
+		if ($datasetId) {
+			$countSql = 'SELECT COUNT(*) FROM omoccurrences o INNER JOIN omoccurdatasetlink dl ON o.occid = dl.occid ' .
+				'WHERE dl.datasetid = ' . $datasetId;
```
Could we use prepared statements here or the `QueryUtil::executeQuery` we should refrain from using direct string concat with sql on new code even though it has been cleaned.

MuchQuak commented on classes/OccurrenceDataset.php:
```diff
@@ -121,208 +129,247 @@ public function getDatasetArr(){
-				'FROM omoccurrences o INNER JOIN omoccurdatasetlink dl ON o.occid = dl.occid '.
-				'WHERE dl.datasetid = '.$datasetId;
+		if ($datasetId) {
+			$sql = 'SELECT o.occid, o.catalognumber, o.occurrenceid ,o.othercatalognumbers, ' .
```
Same as the above comment. Using QueryUtil is probably a good idea here and simple to add. I would only stress about this the code areas you touched because converting everything would be I think unreasonable.

MuchQuak reviewed (changes requested):
Just the small, changes to new/modified sql

Atticus29 committed:
Merge branch 'Development' into dataset-pagination

Atticus29 committed:
change sql queries such that they use QueryUtil::executeQuery

Atticus29 committed:
autoformat classes/OccurrenceDataset.php because it is unlikely to have been touched

Atticus29 commented:
> Just the small, changes to new/modified sql

@MuchQuak should be fixed in e76cc0de0.

MuchQuak reviewed (approved):
Sweet, Changes look good!
